gcc/rust/ast/rust-macro.h

```diff
@@ -479,27 +487,52 @@ class MacroRulesDefinition : public MacroItem
    * I am not aware of the implications of this decision. The rustc spec does
    * mention that using the same parser for macro definitions and invocations
    * is "extremely self-referential and non-intuitive". */
-
-public:
-  std::string as_string () const override;
-
   MacroRulesDefinition (Identifier rule_name, DelimType delim_type,
 			std::vector<MacroRule> rules,
-			std::vector<Attribute> outer_attrs, Location locus)
-    : outer_attrs (std::move (outer_attrs)), rule_name (std::move (rule_name)),
+			std::vector<Attribute> outer_attrs, Location locus,
+			MacroKind kind, Visibility vis)
+    : VisItem (std::move (vis), outer_attrs),
+      outer_attrs (std::move (outer_attrs)), rule_name (std::move (rule_name)),
       delim_type (delim_type), rules (std::move (rules)), locus (locus),
-      associated_transcriber (dummy_builtin), is_builtin_rule (false)
+      associated_transcriber (dummy_builtin), is_builtin_rule (false),
+      kind (kind)
   {}
 
   MacroRulesDefinition (
     Identifier builtin_name, DelimType delim_type,
-    std::function<Fragment (Location, MacroInvocData &)> associated_transcriber)
-    : outer_attrs (std::vector<Attribute> ()), rule_name (builtin_name),
+    std::function<Fragment (Location, MacroInvocData &)> associated_transcriber,
+    MacroKind kind, Visibility vis)
+    : VisItem (std::move (vis), std::vector<Attribute> ()),
+      outer_attrs (std::vector<Attribute> ()), rule_name (builtin_name),
       delim_type (delim_type), rules (std::vector<MacroRule> ()),
       locus (Location ()), associated_transcriber (associated_transcriber),
-      is_builtin_rule (true)
+      is_builtin_rule (true), kind (kind)
   {}
 
+public:
+  std::string as_string () const override;
+
+  static MacroRulesDefinition *mbe (Identifier rule_name, DelimType delim_type,
```

Comment:
We shouldn't return raw pointers like this IMO. Either return `std::unique_ptr<MacroRulesDefinition>`, or have two constructors for `MacroRulesDefinition` and not two static functions which return `MacroRulesDefinition`. We want to enforce memory safety wherever possible and returning raw pointers directly could lead to misuse